engine/apps/mobile_app/tasks.py

```diff
@@ -153,78 +185,256 @@ def _get_fcm_message(alert_group, user, registration_id, critical):
+        return
+
+    # 8 minute window where the notification could be sent (4 mins before or 4 mins after)
+    timing_window_lower = user_notification_timing_preference - NOTIFICATION_TIMING_BUFFER
+    timing_window_upper = user_notification_timing_preference + NOTIFICATION_TIMING_BUFFER
+
+    shift_starts_within_users_notification_timing_preference = _shift_starts_within_range(
+        timing_window_lower, timing_window_upper, seconds_until_shift_starts
+    )
+    shift_starts_within_fifteen_minutes = _shift_starts_within_range(
+        0, FIFTEEN_MINUTES_IN_SECONDS, seconds_until_shift_starts
+    )
+
+    timing_logging_msg = (
+        f"seconds_until_shift_starts: {seconds_until_shift_starts}\n"
+        f"user_notification_timing_preference: {user_notification_timing_preference}\n"
+        f"timing_window_lower: {timing_window_lower}\n"
+        f"timing_window_upper: {timing_window_upper}\n"
+        f"shift_starts_within_users_notification_timing_preference: {shift_starts_within_users_notification_timing_preference}\n"
+        f"shift_starts_within_fifteen_minutes: {shift_starts_within_fifteen_minutes}"
+    )
+
+    if shift_starts_within_users_notification_timing_preference or shift_starts_within_fifteen_minutes:
+        logger.info(f"timing is right to send going oncall push notification\n{timing_logging_msg}")
+        return seconds_until_shift_starts
+    logger.info(f"timing is not right to send going oncall push notification\n{timing_logging_msg}")
+
+
+def _generate_going_oncall_push_notification_cache_key(user_pk: str, schedule_event: ScheduleEvent, time: str) -> str:
+    return f"going_oncall_push_notification:{user_pk}:{schedule_event['shift']['pk']}:{time}"
+
+
+@shared_dedicated_queue_retry_task(autoretry_for=(Exception,), retry_backoff=True, max_retries=MAX_RETRIES)
+def conditionally_send_going_oncall_push_notifications_for_schedule(schedule_pk) -> None:
+    # avoid circular import
+    from apps.mobile_app.models import MobileAppUserSettings
+
+    PUSH_NOTIFICATION_TRACKING_CACHE_KEY_TTL = 80 * 60  # 80 minutes
+    user_cache: typing.Dict[str, User] = {}
+    device_cache: typing.Dict[str, FCMDevice] = {}
+
+    logger.info(f"Start calculate_going_oncall_push_notifications_for_schedule for schedule {schedule_pk}")
+
+    try:
+        schedule: OnCallSchedule = OnCallSchedule.objects.get(pk=schedule_pk)
+    except OnCallSchedule.DoesNotExist:
+        logger.info(f"Tried to notify user about going on-call for non-existing schedule {schedule_pk}")
+        return
+
+    now = timezone.now()
+    current_date_with_hour = now.replace(minute=0, second=0, microsecond=0).strftime("%Y-%m-%d-%Hh")
+    schedule_final_events = schedule.final_events("UTC", now, days=7)
+
+    relevant_cache_keys = [
+        _generate_going_oncall_push_notification_cache_key(user["pk"], schedule_event, current_date_with_hour)
```

Comment:
edge case, but using `current_date_with_hour` couldn't still lead to duplicates? (e.g. my shift starts at 16:02?) wouldn't be enough to use user and event keys? (and caching the key for 1h)